lib/Doctrine/Migrations/Metadata/Storage/TableMetadataStorage.php

```diff
@@ -0,0 +1,141 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Migrations\Metadata\Storage;
+
+use DateTimeImmutable;
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Connections\MasterSlaveConnection;
+use Doctrine\DBAL\Platforms\AbstractPlatform;
+use Doctrine\DBAL\Schema\AbstractSchemaManager;
+use Doctrine\DBAL\Schema\Table;
+use Doctrine\DBAL\Types\Type;
+use Doctrine\Migrations\Metadata\ExecutedMigration;
+use Doctrine\Migrations\Metadata\ExecutedMigrationsSet;
+use Doctrine\Migrations\Version\Direction;
+use Doctrine\Migrations\Version\ExecutionResult;
+use Doctrine\Migrations\Version\Version;
+use InvalidArgumentException;
+use const CASE_LOWER;
+use function array_change_key_case;
+use function intval;
+use function sprintf;
+use function strtolower;
+
+final class TableMetadataStorage implements MetadataStorage
+{
+    /** @var Connection */
+    private $connection;
+
+    /** @var AbstractSchemaManager */
+    private $schemaManager;
+
+    /** @var AbstractPlatform */
+    private $platform;
+
+    /** @var TableMetadataStorageConfiguration */
+    private $configuration;
+
+    public function __construct(Connection $connection, ?MetadataStorageConfigration $configuration = null)
+    {
+        $this->connection    = $connection;
+        $this->schemaManager = $connection->getSchemaManager();
+        $this->platform      = $connection->getDatabasePlatform();
+
+        if ($configuration !== null && ! ($configuration instanceof TableMetadataStorageConfiguration)) {
```

Comment:
We could change the type hint to `TableMetadataStorageConfiguration`, as widening it to `MetadataStorageConfiguration` later (once we support different configurations) is perfectly acceptable without a BC break.

Reply:
This is nothing but a side-effect of https://github.com/doctrine/migrations/pull/858#discussion_r344870590, though. Solving the root cause should allow us to remove this condition.